modules/xml/filterx-parse-xml.c

```diff
@@ -0,0 +1,702 @@
+{
+  GQueue *obj_stack = (GQueue *) user_data;
+  XmlElemContext *elem_context = g_queue_peek_head(obj_stack);
+  const gchar *element_name = g_markup_parse_context_get_element(context);
+
+  gsize stripped_text_len;
+  gchar *stripped_text = _strip(text, text_len, &stripped_text_len);
+  if (!stripped_text)
+    {
+      /*
+       * We already stored an empty value in _prepare_elem() and _create_object_for_new_elem().
+       * There's nothing to do here.
+       */
+      return;
+    }
+
+  if (filterx_object_is_type(elem_context->current_obj, &FILTERX_TYPE_NAME(string)))
+    {
+      _replace_string_text(elem_context, element_name, stripped_text, stripped_text_len, error);
+      goto exit;
+    }
+
+  if (filterx_object_is_type(elem_context->current_obj, &FILTERX_TYPE_NAME(dict)))
+    {
+      _add_text_to_dict(elem_context, stripped_text, stripped_text_len, error);
+      goto exit;
+    }
+
+  g_assert_not_reached();
+
+exit:
+  g_free(stripped_text);
+}
+
+static gboolean
+_validate_fillable(FilterXGeneratorFunctionParseXml *self, FilterXObject *fillable)
+{
+  if (!filterx_object_is_type(fillable, &FILTERX_TYPE_NAME(dict)))
+    {
+      filterx_eval_push_error_info("fillable must be dict", &self->super.super.super,
+                                   g_strdup_printf("got %s instead", fillable->type->name), TRUE);
+      return FALSE;
+    }
+  return TRUE;
+}
+
+static const gchar *
+_extract_raw_xml(FilterXGeneratorFunctionParseXml *self, FilterXObject *xml_obj, gsize *len)
+{
+  const gchar *raw_xml;
+  if (!filterx_object_extract_string(xml_obj, &raw_xml, len))
+    {
+      filterx_eval_push_error_info("input must be string", &self->super.super.super,
+                                   g_strdup_printf("got %s instead", xml_obj->type->name), TRUE);
+      filterx_object_unref(xml_obj);
+      return NULL;
+    }
+
+  return raw_xml;
+}
+
+static gboolean
+_parse(FilterXGeneratorFunctionParseXml *self, const gchar *raw_xml, gsize raw_xml_len, FilterXObject *fillable)
+{
+  static GMarkupParser scanner_callbacks =
+  {
+    .start_element = _start_elem_cb,
+    .end_element = _end_elem_cb,
+    .text = _text_cb,
+  };
+
+  GQueue *obj_stack = g_queue_new();
+  XmlElemContext *root_elem_context = _elem_context_new(NULL, fillable);
+  g_queue_push_head(obj_stack, root_elem_context);
+  GMarkupParseContext *context = g_markup_parse_context_new(&scanner_callbacks, 0, obj_stack, NULL);
+
+  GError *error = NULL;
+  gboolean success = g_markup_parse_context_parse(context, raw_xml, raw_xml_len, &error) &&
+                     g_markup_parse_context_end_parse(context, &error);
+  if (!success)
+    {
+      gchar *error_info = g_strdup(error ? error->message : "unknown error");
+      filterx_eval_push_error_info("failed to parse xml", &self->super.super.super, error_info, TRUE);
+      if (error)
+        g_error_free(error);
+      goto exit;
+    }
+
+exit:
+  _elem_context_free(root_elem_context);
+  g_queue_free(obj_stack);
```

Comment:
In case of errors, the XmlElemContexts in the queue need to be freed (unclosed tags).

Reply:
Thanks, fixed.